libmultilabel/model.py

```diff
@@ -59,40 +60,55 @@ def training_step(self, batch, batch_idx):
         return loss
 
     def validation_step(self, batch, batch_idx):
-        loss, pred_logits = self.shared_step(batch)
-        return {'loss': loss.item(),
-                'pred_scores': torch.sigmoid(pred_logits).detach().cpu().numpy(),
-                'target': batch['label'].detach().cpu().numpy()}
+        self._shared_eval_step(batch, datch_idx)
+
+    def validation_step_end(self, batch_parts):
+        self._shared_eval_step_end(batch_parts)
 
     def validation_epoch_end(self, step_outputs):
-        eval_metric = self.evaluate(step_outputs, 'val')
-        return eval_metric
+        return self._shared_eval_epoch_end(step_outputs, 'val')
 
     def test_step(self, batch, batch_idx):
-        return self.validation_step(batch, batch_idx)
+        return self._shared_eval_step(batch, batch_idx)
+
+    def test_step_end(self, batch_parts):
+        self._shared_eval_step_end(batch_parts)
 
     def test_epoch_end(self, step_outputs):
-        eval_metric = self.evaluate(step_outputs, 'test')
-        self.test_results = eval_metric
-        return eval_metric
-
-    def evaluate(self, step_outputs, split):
-        eval_metric = MultiLabelMetrics(self.config)
-        for step_output in step_outputs:
-            eval_metric.add_values(y_pred=step_output['pred_scores'],
-                                   y_true=step_output['target'])
-        metric_dict = eval_metric.get_metric_dict()
+        return self._shared_eval_epoch_end(step_outputs, 'test')
+
+    def _shared_eval_step(self, batch, batch_idx):
+        loss, pred_logits = self.shared_step(batch)
+        return {'loss': loss.item(),
+                'pred_scores': torch.sigmoid(pred_logits).detach().cpu().numpy(),
+                'target': batch['label'].detach().cpu().numpy()}
+
+    def _shared_eval_step_end(self, batch_parts):
+        pred_scores = np.vstack(batch_parts['pred_scores'])
+        target = np.vstack(batch_parts['target'])
+        self.eval_metric.update(target, pred_scores)
+
+    def _shared_eval_epoch_end(self, step_outputs, split):
+        metric_dict = self.eval_metric.get_metric_dict()
         self.log_dict(metric_dict)
         dump_log(config=self.config, metrics=metric_dict, split=split)
 
-        self.print(f'\n====== {split.upper()} dataset evaluation result =======')
-        self.print(eval_metric)
-        return eval_metric
+        if not self.config.silent and (not self.trainer or self.trainer.is_global_zero):
+            print(f'====== {split} dataset evaluation result =======')
+            print(self.eval_metric)
+            print()
+        self.eval_metric.reset()
+        return metric_dict
+
+    def predict_step(self, batch, batch_idx, dataloader_idx):
```

Comment:
where is the function used now? I don't think it's called?

Reply:
It is used by trainer.predict(), which is in the line 138 of main.py.